### PR TITLE
Debug of numeric 3-state errors in clustered CPMG analysis

### PR DESCRIPTION
**Pull request by ilsacripante**
I believe I could find some more errors in the implementation of the MMQ NS 3 site exchange models, now calculations seem to proceed. 

Dear @edward-dauvergne : I can send a full dataset for the data I am using to test the models but it is still unpublished so I would like you not to make them available to the public before I tell you. my email is luca.codutti > oci.uni-hannover.de please contact me if you wish to receive it

Error tracing started as MPI would break calculations with a general 
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD

this was due to 2 mistakes in the vector definition of some NS-MQ parameters in /specific_analyses/relax_disp/parameters.py in lines 783 and 919

the master version of relax seems to work only because in line 783 the dw/H_AC params are forgotten and "de facto" never optimized. After fixing however some new errors arise

1st error in grid search
relax gives RelaxWarning error: invalid value encountered in true_divide during grid search. 
nan data give inf  values during simplex minimization which break  the minimization.

The trace back brings you to the ns_mmq_3site.py function, and it turns out that the error is straightforward: 
pB_pC is defined in  relax/lib/dispersion/ns_mmq_3site.py as 

`   pB_pC = pB + pC
    k_BA = pA * kex_AB / pA_pB
    k_AB = pB * kex_AB / pA_pB
    k_CB = pB * kex_BC / pB_pC
    k_BC = pC * kex_BC / pB_pC`

As pC is defined as 1 -pA -pB, you get a divide by 0 error 3 lines later if pA is set to 1 and pB to 0 in the grid search. For this reason I added a pB_lower and a pA_upper function in /specific_analyses/relax_disp/parameter_object.py which specifically sets pB_lower  to 0.01 if 3site exchange models are used and pA_upper to 0.99. This is a workaround, if you guys have a better solution I would be glad.
      
2nd error in monte carlo error determination : pC_sim is not found, this is due to typo  in line 813 of the same file. AFAIK pC is never simulated but only given as difference between 1 pA and pB, am I right? so after substituting with None the sim attribute, the error is not there anymore

### CODE REVIEW

File: auto_analyses/relax_disp.py [hunk 1]
```diff
@@ -208,7 +208,9 @@ def check_numpy_less_1_8_and_numerical_model(self):
         """Check for numerical model using numpy version under 1.8.  This will result in slow "for loop" calculation through data, making the analysis 5-6 times slower."""
 
         # Some warning for the user if the pure numeric solution is selected.
-        if float(version.version[:3]) < 1.8:
```
Comment (tlinnet) [inline auto_analyses/relax_disp.py:211]: Hi @ilsacripante . What is the reason for this change?

File: devel_scripts/code_validator [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#! /usr/bin/python
+#! /usr/bin/env python
```
Comment (tlinnet) [inline devel_scripts/code_validator:1]: Good change, but belongs to another PR

File: devel_scripts/fsfcv [hunk 3]
```diff
@@ -1,4 +1,4 @@
-#! /usr/bin/env python3
+#! /usr/bin/env python
```
Comment (tlinnet) [inline devel_scripts/fsfcv:1]: This is a devel script. Are you sure that Python 2.7 can handle this?

And this belongs to another PR.

File: devel_scripts/openmpi_test_install_tcsh.sh [hunk 4]
```diff
@@ -41,12 +41,12 @@ echo "> mpirun --report-bindings -np 2 echo $A hello world $A"; mpirun --report-
 
 echo "> mpirun --report-bindings -np 2 python -c $A print "\""hello"\"" $A"; mpirun --report-bindings -np 2 python -c "print $A hello $A"; echo ""
 
-echo "> mpirun --report-bindings -np 2 python --version"; mpirun --report-bindings -np 2 python --version; echo ""
+echo "> mpirun --report-bindings -np 2 python --version"; mpirun --report-bindings -np 2 python3 --version; echo ""
```
Comment (tlinnet) [inline devel_scripts/openmpi_test_install_tcsh.sh:44]: These changes belongs to another PR.

And python3 should be python.
It has to be the user who setup which version of python is callen.

File: docs/devel/svn2git_migration/dups.py [hunk 5]
```diff
@@ -1,4 +1,4 @@
-#! /usr/bin/env python3
```
Comment (tlinnet) [inline docs/devel/svn2git_migration/dups.py:1]: This is a devel script. Are you sure that Python 2.7 can handle this?

And this belongs to another PR.

File: docs/devel/svn2git_migration/git_cleanup.py [hunk 6]
```diff
@@ -1,4 +1,4 @@
-#! /usr/bin/env python3
```
Comment (tlinnet) [inline docs/devel/svn2git_migration/git_cleanup.py:1]: This is a devel script. Are you sure that Python 2.7 can handle this?

And this belongs to another PR.

File: docs/devel/svn2git_migration/lost_commits.py [hunk 7]
```diff
@@ -1,4 +1,4 @@
-#! /usr/bin/env python3
```
Comment (tlinnet) [inline docs/devel/svn2git_migration/lost_commits.py:1]: This is a devel script. Are you sure that Python 2.7 can handle this?

And this belongs to another PR.

File: docs/devel/svn2git_migration/merge_reconstruction.py [hunk 8]
```diff
@@ -1,4 +1,4 @@
-#! /usr/bin/env python3
```
Comment (tlinnet) [inline docs/devel/svn2git_migration/merge_reconstruction.py:1]: This is a devel script. Are you sure that Python 2.7 can handle this?

And this belongs to another PR.

File: lib/dispersion/mmq_cr72.py [hunk 9]
```diff
@@ -81,6 +83,17 @@ def r2eff_mmq_cr72(r20=None, pA=None, dw=None, dwH=None, kex=None, cpmg_frqs=Non
     @type back_calc:        numpy float array of rank [NS][NM][NO][ND]
     """
 
+    #parameters must be assigned. If none present raise exception
+    for x in  [r20,pA,dw,dwH,kex]:
```
Comment (tlinnet) [inline lib/dispersion/mmq_cr72.py]: 1 space " " after comma ","

File: lib/dispersion/ns_mmq_2site.py [hunk 10]
```diff
@@ -333,6 +358,16 @@ def r2eff_ns_mmq_2site_sq_dq_zq(M0=None, F_vector=array([1, 0], float64), R20A=N
     @keyword power:         The matrix exponential power array.
     @type power:            numpy int array of rank [NS][NM][NO][ND]
     """
+    for x in [R20A,R20B,dw,dwH,pA,kex]:
```
Comment (tlinnet) [inline lib/dispersion/ns_mmq_2site.py]: 1 space " " after comma ","

File: lib/dispersion/ns_mmq_3site.py [hunk 11]
```diff
@@ -167,6 +169,16 @@ def rmmq_3site_rankN(R20A=None, R20B=None, R20C=None, dw_AB=None, dw_AC=None, k_
     #matrix[2, 0] = k_AC
     #matrix[2, 1] = k_BC
     #matrix[2, 2] = -k_CB - k_CA + 1.j*dw_AC - R20C
+    for x in [R20A,R20B,R20C,dw_AB,dw_AC,k_AB,k_BA,k_BC,k_CB,k_AC,k_CA]:
```
Comment (tlinnet) [inline lib/dispersion/ns_mmq_3site.py]: 1 space " " after comma ","

File: lib/dispersion/ns_mmq_3site.py [hunk 12]
```diff
@@ -207,6 +219,13 @@ def rmmq_3site_rankN(R20A=None, R20B=None, R20C=None, dw_AB=None, dw_AC=None, k_
             + m_k_AB_tcp + m_k_BA_tcp + m_k_BC_tcp
             + m_k_CB_tcp + m_k_AC_tcp + m_k_CA_tcp)
 
+    #print(matrix)
```
Comment (tlinnet) [inline lib/dispersion/ns_mmq_3site.py]: remove comment code

File: lib/dispersion/ns_mmq_3site.py [hunk 13]
```diff
@@ -264,6 +283,17 @@ def r2eff_ns_mmq_3site_mq(M0=None, F_vector=array([1, 0, 0], float64), R20A=None
     @type power:            numpy int array of rank [NS][NM][NO][ND]
     """
 
+    for x in [R20A,R20B,R20C,pA,pB,dw_AB,dw_BC,dwH_AB,dwH_BC,kex_AB,kex_BC,kex_AC]:
```
Comment (tlinnet) [inline lib/dispersion/ns_mmq_3site.py]: 1 space " " after comma ","

File: lib/dispersion/ns_mmq_3site.py [hunk 14]
```diff
@@ -448,6 +478,16 @@ def r2eff_ns_mmq_3site_sq_dq_zq(M0=None, F_vector=array([1, 0, 0], float64), R20
     @keyword power:         The matrix exponential power array.
     @type power:            numpy int array of rank [NS][NM][NO][ND]
     """
+    for x in [R20A,R20B,R20C,pA,pB,dw_AB,dw_BC,dwH_AB,dwH_BC,kex_AB,kex_BC,kex_AC]:
```
Comment (tlinnet) [inline lib/dispersion/ns_mmq_3site.py]: 1 space " " after comma ","

**Review by tlinnet — COMMENTED** [review]
This looks great.

1) Can you change the title of the PR? What is this PR for?
2) Can you separate the PR into 2 different PR?
One PR for fixing the model and one PR for changing all the "python" suggestions
3) Can you create a Systemtests that demonstrates the failure of the model before the fix?

**Comment by edward-dauvergne** [comment]
Note that I am setting up the relax mailing lists at SourceForge:  https://sourceforge.net/p/nmr-relax/mailman/.  The relax development mailing list should be used for most discussions.  The primary git repository is also at SourceForge and the ones on GitHub, GitLab, and Bitbucket are mirrors.  Where forks are located are not important.  Also note that GitHub pull requests will not be used as these are not real git merge requests.  Ideally development would be:  You would create a new feature branch, develop the code there in a number of commits, and then email the development mailing list with a URL to the repository and the branch name.  That way the commits can be reviewed and a proper git merge performed.

File: devel_scripts/code_validator (hunk 2 above)
Comment (edward-dauvergne) [reply devel_scripts/code_validator:1]: This is reasonable.  Note that this is the script that you should use to check your code changes!  And, yes, this should be separate.  In a separate branch and in a separate commit.

File: devel_scripts/fsfcv (hunk 3 above)
Comment (edward-dauvergne) [reply devel_scripts/fsfcv:1]: This change is no good.  I wrote this script solely for Python3!  Supporting Python2 in this script is not desirable.

File: docs/devel/svn2git_migration/dups.py (hunk 5 above)
Comment (edward-dauvergne) [reply docs/devel/svn2git_migration/dups.py:1]: The svn2git_migration scripts are an archive of the scripts used to convert the relax Subversion repository to git.  Why are these being changed?

File: lib/dispersion/ns_mmq_2site.py (hunk 10 above)
Comment (edward-dauvergne) [reply lib/dispersion/ns_mmq_2site.py]: This is what the devel_scripts/code_validator script will highlight as well!

**Comment by edward-dauvergne** [comment]
Also, I am in the process of rewriting the development chapter of the relax manual at the moment (http://www.nmr-relax.com/manual/relax_development.html).  I will include all of this information there.  If you haven't read that chapter, you really should.  Just note that in the current version, I am changing all Gna! references to SourceForge, and changing all svn information to git.  Some sections are also being deleted (for example the info about sending patches).  If you can update your master branch and run:

```
$ scons user_manual_pdf
```

Then you will have the chapter half rewritten.

File: auto_analyses/relax_disp.py (hunk 1 above)
Comment (ilsacripante) [reply auto_analyses/relax_disp.py:211]: sorry @tlinnet I just got back from other duties and I am ready to respond to your requests. 
This is just an annoyance: the original code always evaluates numpy version > 1.10 as older than 1.8, it bugged me a bit and I fixed it by changing the lower accepted version number to 1.08 as 1.10 > 1.08 but not 1.80

File: devel_scripts/fsfcv (hunk 3 above)
Comment (ilsacripante) [reply devel_scripts/fsfcv:1]: sorry about that, during my peregrination through the code I wanted to be sure there were no calls to my other python version and this bit remained altered but must stay as in the master

File: devel_scripts/openmpi_test_install_tcsh.sh (hunk 4 above)
Comment (ilsacripante) [reply devel_scripts/openmpi_test_install_tcsh.sh:44]: sorry about that, during my peregrination through the code I wanted to be sure there were no calls to my other python version and this bit remained altered but must stay as in the master

File: docs/devel/svn2git_migration/dups.py (hunk 5 above)
Comment (ilsacripante) [reply docs/devel/svn2git_migration/dups.py:1]: sorry about that, during my peregrination through the code I wanted to be sure there were no calls to my other python version and this bit remained altered but must stay as in the master

File: auto_analyses/relax_disp.py (hunk 1 above)
Comment (edward-dauvergne) [reply auto_analyses/relax_disp.py:211]: Please use the relax dep_check.version_comparison() function here!  I wrote this function specifically for comparing such version numbers.

**Comment by edward-dauvergne** [comment]
This has been fixed upstream.
